client/src/unifyfs_fid.c

```diff
@@ -371,24 +371,89 @@ int unifyfs_fid_delete(unifyfs_client* client,
     return UNIFYFS_SUCCESS;
 }
 
-/* opens a new file id with specified path, access flags, and permissions,
- * fills outfid with file id and outpos with position for current file pointer,
- * returns UNIFYFS error code
+/* Given a file name, allocate a gfid entry on the server for the file. */
+int unifyfs_gfid_create(
+    unifyfs_client* client,
+    const char* path, /* path of file to be created */
+    int exclusive,    /* return EEXIST if gfid already exists */
+    int private,      /* create node-local (non-shared) file */
+    int truncate,     /* truncate file to 0 length if already exists */
+    mode_t mode)      /* mode bits as from open(2) */
+{
+    /* check that pathname is within bounds */
+    size_t pathlen = strlen(path) + 1;
+    if (pathlen > UNIFYFS_MAX_FILENAME) {
+        return ENAMETOOLONG;
+    }
+
+    int gfid = unifyfs_generate_gfid(path);
+
+    /* initialize file attributes */
+    unifyfs_file_attr_t gfattr = { 0, };
+    unifyfs_file_attr_set_invalid(&gfattr);
+
+    gfattr.gfid = gfid;
+    gfattr.size = 0;
+    gfattr.mode = mode;
+    gfattr.is_laminated = 0;
+    gfattr.is_shared = !private;
+    gfattr.filename = (char*) path;
+
+    /* use client user/group */
+    gfattr.uid = getuid();
+    gfattr.gid = getgid();
+
+    /* use current time for atime/mtime/ctime */
+    struct timespec tp = {0};
+    clock_gettime(CLOCK_REALTIME, &tp);
+    gfattr.atime = tp;
+    gfattr.mtime = tp;
+    gfattr.ctime = tp;
+
+    /* TODO: process O_CREAT, O_EXCL, and O_TRUNC at server */
+
+    /* Attempt to create gfid entry for file on server */
+    unifyfs_file_attr_op_e op = UNIFYFS_FILE_ATTR_OP_CREATE;
+    int ret = unifyfs_set_global_file_meta(client, gfid, op, &gfattr);
+    if (ret == EEXIST) {
+        LOGINFO("Attempt to create existing file %s (gfid:%d)", path, gfid);
+        if (!exclusive) {
+            ret = UNIFYFS_SUCCESS;
+        } else {
+            LOGERR("Failed create of existing private/exclusive file %s",
+                   path);
+        }
+    }
+    if (ret != UNIFYFS_SUCCESS) {
+        return ret;
+    }
+
+#if 0
```

Comment:
I suggest removing this if0 block since like you mention in the comment, this is better handled by the server.

Reply:
Good point.  I dropped the ``#if 0`` code.  I left the TODO comment. in there as a reminder for future work on the server, since it helps explain why we pass ``truncate`` as an input parameter but don't use it.

Reply:
After chatting in the meeting, I'll drop the TODO, as well.